### ISSUE TITLE
Board.legal() fails after Board.null_move()

### ISSUE THREAD

**Issue by mattbruv** [jordanbray/chess#11]
The Board.legal() function fails at recognizing legal moves/pins after Board.null_move() is called. Take the [following example position](https://lichess.org/analysis/4k3/3r4/8/8/8/8/3R4/3K4_w_-_-). The positions start exactly the same way except for the side to move. In the white example, it considers Ra2 illegal as it should. However in the exact same position, with white to move (after a null move from black), it considers Ra2 legal.

```
extern crate chess;

use chess::*;

fn main() {

	let pos_white     = Board::from_fen("4k3/3r4/8/8/8/8/3R4/3K4 w - - 0 1".to_owned()).unwrap();
	let mut pos_black = Board::from_fen("4k3/3r4/8/8/8/8/3R4/3K4 b - - 0 1".to_owned()).unwrap();

	let testmove = ChessMove::new(
		Square::from_string("d2".to_owned()).unwrap(),
		Square::from_string("a2".to_owned()).unwrap(),
		None);

	// Ra2 should fail because of the pin
	assert!(pos_white.legal(testmove) == false);

	// Same exact position, but black has the move.
	// Let's change the side to move to white via null_move
	pos_black = pos_black.null_move().unwrap();

	// Make sure white is the side to move
	assert!(pos_black.side_to_move() == Color::White);

	// Ra2 should be illegal, because the position is exactly the same as pos_white
	assert!(pos_black.legal(testmove) == false); // assertion failed
}
```

**Comment by jordanbray** [comment]
Clearly, the null_move() function is not updating the board state well enough to do proper legality checks. Let me look into this over the weekend and confirm and fix this issue.

**Comment by mattbruv** [comment]
Sounds good. Thanks a lot for making this library. It takes out so much of the heavy work of a chess engine and has been a pleasure to use in my project so far.

**Comment by jordanbray** [comment]
No problem.  Thanks for using the library.  The stated goal of this project is to prevent every chess engine developer from writing their own move generator (I've been there myself.)  I'm happy to take the burden off of other people in a useful way.

**Comment by jordanbray** [comment]
Sorry for forgetting about this one.  The patch looks good, I tested it as well.  It is merged and version 0.4.2 is released.
